### PR TITLE
Add nft index

### LINKED ISSUE #919: Implement an index which tracks the last UTXO of each NFT

### LINKED ISSUE #920: Show last UTXO of an NFT through RPC command

### PR DESCRIPTION
**Pull request by mxaddict**
closes #919
closes #920

### CODE REVIEW

**Comment by navbuilder** [comment]
A new build of 3fe3737378e6bac899a72804f79271d134b6809c has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 3fe3737378e6bac899a72804f79271d134b6809c has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by mxaddict** [comment]
image

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -6028,15 +6036,27 @@ UniValue listtokens(const UniValue& params, bool fHelp)
...
                     n.pushKV("metadata", it_.second);
                     int64_t tempBalance = pwalletMain->GetPrivateBalance(TokenId(it->first, it_.first));
                     n.pushKV("balance", std::to_string(tempBalance));
                     balance += tempBalance;
+
+                    std::vector<CTxOut> utxos;
+                    if (fWithUtxo && GetNftUnspentIndex(it->first, utxos)) {
+                        auto txout = utxos[utxos.size() - 1];
+                        UniValue utxo(UniValue::VOBJ);
+                        utxo.pushKV("hash", txout.GetHash().ToString());
```
Comment (aguycalled) [inline src/wallet/rpcwallet.cpp:6052]: it should show txout.hash.ToString() and txout.n

**Comment by navbuilder** [comment]
A new build of b3d81e47ab587c9af3a61d29e4b5dc515d2f44c6 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by mxaddict** [comment]
image

**Comment by mxaddict** [comment]
@chasingkirkjufell @sakdeniz once build for 4afb994 is done, I think you should give this a test :)

**Comment by mxaddict** [comment]
@aguycalled I've also added the (with_utxo) param to `gettoken` rpc

**Comment by navbuilder** [comment]
A new build of 532baf159139f3c6f8c27a91c35c53ae904a0279 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 82f14bd7bf4f45e21b39600d9be335eff3f44bd2 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by chasingkirkjufell** [comment]
from discord conversations it sounds like you're still working on this?

**Comment by mxaddict** [comment]
Yes, I'm moving some logic into stateview

**Review by aguycalled — COMMENTED** [review]
changes look good, will test tomorrow and report back

**Comment by mxaddict** [comment]
Thanks

**Comment by navbuilder** [comment]
A new build of e2df75faa927ab1713c6b04d87275011e84d0b6c has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by aguycalled** [comment]
the daemon does not check for discrepancies between `-nftindex` and the local db state (for example, if the local db has a nft index but `-nftindex=0` was specified)

**Comment by aguycalled** [comment]
image

Invalidating a block does not return to the previous utxo

**Comment by mxaddict** [comment]
@aguycalled I'm checking on this now

**Comment by mxaddict** [comment]
@aguycalled what did you mean by this:
```
the daemon does not check for discrepancies between -nftindex and the local db state (for example, if the local db has a nft index but -nftindex=0 was specified)
```

**Comment by navbuilder** [comment]
A new build of 3183371b29daafc213f44f54eec973ebb950d91c has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 7681f54d3a786f142db9990604c8a690374ee2a7 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of cc4634ce5bd9bc7f6d3c8ecb4e45036bd670a1b3 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 7c18569130b61d036fb1f3cea1527156fc906358 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 32f1884aa322bd0e97ba64f49e58a8e0291de165 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Review by aguycalled — APPROVED** [review]

**Comment by navbuilder** [comment]
A new build of 3bcfbb579a1222996574c087624efb5e2b574bf9 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 9dbbf583cf117a125fcd5f0a5898f2e3e4109be2 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of 6e1f7edbe92a0779d5cbc377a13b07d1534c87d0 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Comment by navbuilder** [comment]
A new build of bb1adde3e9933af2d79b11a4b602abb8883a9ba5 has completed succesfully!
Binaries available at https://build.nav.community/binaries/mxaddict:add-nft-index

**Review by chasingkirkjufell — APPROVED** [review]
looks good to me. thanks for cleaning up the stressor. you added the checks i wanted to add as well! sorry for the delays.

**Pull request merged**